operators/test/e2e/es/reversal_test.go

```diff
@@ -0,0 +1,96 @@
+		WithESDataNodes(1, elasticsearch.DefaultResources).
+		WithESMasterNodes(1, elasticsearch.DefaultResources)
+
+	bogus := b.WithAdditionalConfig(map[string]map[string]interface{}{
+		"data": {
+			"this leads": "to a bootlooping instance",
+		},
+	})
+
+	state := elasticsearch.NewMutationReversalTestState(b.Elasticsearch)
+	test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{bogus}, state)
+}
+
+func TestReversalRiskyMasterDownscale(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-non-ha-downscale-reversal").
+		WithESMasterDataNodes(2, elasticsearch.DefaultResources)
+	down := b.WithNoESTopology().WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	state := elasticsearch.NewMutationReversalTestState(b.Elasticsearch)
+	test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{down}, state)
+}
+
+// TODO we validate a downscale to replica=0 not sure if we should allow it
+func TestReversalSingleMasterDownscale(t *testing.T) {
+	t.Skip()
+	b := elasticsearch.NewBuilder("test-non-ha-downscale-reversal").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+	down := b.WithNoESTopology().WithESMasterDataNodes(0, elasticsearch.DefaultResources)
+
+	state := elasticsearch.NewMutationReversalTestState(b.Elasticsearch)
+	test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{down}, state)
+}
+
+func TestReversalStatefulSetRename(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-sset-rename-reversal").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	copy := b.Elasticsearch.Spec.Nodes[0]
+	copy.Name = "other"
+	renamed := b.WithNoESTopology().WithNodeSpec(copy)
+
+	state := elasticsearch.NewMutationReversalTestState(b.Elasticsearch)
+	test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{renamed}, state)
+}
+
+func TestRiskyMasterReconfiguration(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-sset-reconfig-reversal").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithNodeSpec(v1alpha1.NodeSpec{
+			Name:      "other-master",
+			NodeCount: 1,
+			Config: &common.Config{
+				Data: map[string]interface{}{
+					v1alpha1.NodeMaster: true,
+					v1alpha1.NodeData:   true,
+				},
+			},
+			PodTemplate: elasticsearch.ESPodTemplate(elasticsearch.DefaultResources),
+		})
+
+	// this currently breaks the cluster (something we might fix in the future at which point this just tests a temp downscale)
+	noMasterMaster := b.WithNoESTopology().WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithNodeSpec(v1alpha1.NodeSpec{
+			Name:      "other-master",
+			NodeCount: 1,
+			Config: &common.Config{
+				Data: map[string]interface{}{
+					v1alpha1.NodeMaster: false,
+					v1alpha1.NodeData:   true,
+				},
+			},
+			PodTemplate: elasticsearch.ESPodTemplate(elasticsearch.DefaultResources),
+		})
+
+	state := elasticsearch.NewMutationReversalTestState(b.Elasticsearch)
+	test.RunMutationReversal(t, []test.Builder{b}, []test.Builder{noMasterMaster}, state)
```

Comment:
I feel like it's a bit strange to build and propagate the `state` here. Could `test.RunMutationReversal` just build it directly (internally) based on the creation builders?

Reply:
My thinking was that we could have similar tests for APM and Kibana and then the state would be specific to the resource, but I agree that that is maybe a bit premature. I am not sure these back and forth scenarios make sense with the simpler resource types.